File: UI/Ecommerce.UI/Services/CartService.cs
using Ecommerce.Shared.Dtos;
using Ecommerce.Shared.Requests.CartItem;
using Ecommerce.Shared.Responses.CartItem;
using Ecommerce.UI.Contracts;
using Ecommerce.UI.Contracts.Refit;
using Newtonsoft.Json;
using Refit;
using System;
using System.Threading.Tasks;

namespace Ecommerce.UI.Services
{
	public class CartService : ICartService
	{
		private readonly ICartApiService _cartApiService;

		public CartService(ICartApiService cartApiService)
		{
			this._cartApiService = cartApiService;
		}
		
		public async Task<CreateCartItemResponse> AddItemToCart(CartItemDto cartItem)
		{
			ApiResponse<CreateCartItemResponse> response = await this._cartApiService.CreateCartItem(new CreateCartItemApiRequest { CartItemToCreate = cartItem });
			
			if (response.IsSuccessStatusCode)
			{
				return response.Content;
			}
			
			return string.IsNullOrEmpty(response.Error.Content) ? 
				new CreateCartItemResponse { Success = false, Message = "Unexpected Error Occurred" } 
				: JsonConvert.DeserializeObject<CreateCartItemResponse>(response.Error.Content)!;
		}

		public async Task<DeleteCartItemResponse> RemoveItemFromCart(CartItemDto cartItem)
		{
			ApiResponse<DeleteCartItemResponse> response = await this._cartApiService.DeleteCartItem(new DeleteCartItemApiRequest { CartItemToDelete = cartItem });
			
			if (response.IsSuccessStatusCode)
			{
				return response.Content;
			}
			
			return string.IsNullOrEmpty(response.Error.Content) ? 
				new DeleteCartItemResponse { Success = false, Message = "Unexpected Error Occurred" } 
				: JsonConvert.DeserializeObject<DeleteCartItemResponse>(response.Error.Content)!;
		}

		public async Task<DeleteUserCartItemsResponse> ClearCart(Guid userId)
		{
			ApiResponse<DeleteUserCartItemsResponse> response = await this._cartApiService.DeleteAllCartItems(new DeleteUserCartItemsApiRequest { UserId = userId });
			
			if (response.IsSuccessStatusCode)
			{
				return response.Content;
			}
			
			return string.IsNullOrEmpty(response.Error.Content) ? 
				new DeleteUserCartItemsResponse { Success = false, Message = "Unexpected Error Occurred" } 
				: JsonConvert.DeserializeObject<DeleteUserCartItemsResponse>(response.Error.Content)!;
		}

		public async Task<UpdateCartItemResponse> UpdateItemInCart(CartItemDto cartItem)
		{
			ApiResponse<UpdateCartItemResponse> response = await this._cartApiService.UpdateCartItem(new UpdateCartItemApiRequest { CartItemToUpdate = cartItem });
			
			if (response.IsSuccessStatusCode)
			{
				return response.Content;
			}
			
			return string.IsNullOrEmpty(response.Error.Content) ? 
				new UpdateCartItemResponse { Success = false, Message = "Unexpected Error Occurred" } 
				: JsonConvert.DeserializeObject<UpdateCartItemResponse>(response.Error.Content)!;
		}

		public async Task<GetUserCartItemsResponse> GetItemsInCart(Guid userId)
		{
			ApiResponse<GetUserCartItemsResponse> response = await this._cartApiService.GetUserCartItems(new GetUserCartItemsApiRequest { UserId = userId });
			
			if (response.IsSuccessStatusCode)
			{
				return response.Content;
			}
			
			return string.IsNullOrEmpty(response.Error.Content) ? 
				new GetUserCartItemsResponse { Success = false, Message = "Unexpected Error Occurred" } 
				: JsonConvert.DeserializeObject<GetUserCartItemsResponse>(response.Error.Content)!;
		}
	}
}

File: UI/ecommerce-ui/src/services/CartService.ts
import {CartItem} from "../models/CartItem.ts";
import axiosInstance from "../AxiosInstance.ts";

const addItemToCart = async (cartItem: CartItem) => {
    try {
        const response = await axiosInstance.post(`/api/cartitem/create`, {
            CartItemToCreate: cartItem
        });

        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }

        return {
            success: false,
            message: response.data.message || "Unexpected Error Occurred",
            validationErrors: []
        };
    } catch (error) {
        return {
            success: false,
            message: "Unexpected Error Occurred",
            validationErrors: []
        };
    }
}

const removeItemFromCart = async (cartItem: CartItem) => {
    try {
        const response = await axiosInstance.delete(`/api/cartitem/delete`, {
            data: { CartItemToDelete: cartItem }
        });

        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }

        return {
            success: false,
            message: response.data.message || "Unexpected Error Occurred",
            validationErrors: []
        };
    } catch (error) {
        return {
            success: false,
            message: "Unexpected Error Occurred",
            validationErrors: []
        };
    }
}

const clearCart = async (userId: string) => {
    try {
        const response = await axiosInstance.delete(`/api/cartitem/user/delete`, {
            data: { UserId: userId }
        });

        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }

        return {
            success: false,
            message: response.data.message || "Unexpected Error Occurred",
            validationErrors: []
        };
    } catch (error) {
        return {
            success: false,
            message: "Unexpected Error Occurred",
            validationErrors: []
        };
    }
}

const updateItemInCart = async (cartItem: CartItem) => {
    try {
        const response = await axiosInstance.put(`/api/cartitem/update`, {
            CartItemToUpdate: cartItem
        });

        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }

        return {
            success: false,
            message: response.data.message || "Unexpected Error Occurred",
            validationErrors: []
        };
    } catch (error) {
        return {
            success: false,
            message: "Unexpected Error Occurred",
            validationErrors: []
        };
    }
}

const getItemsInCart = async (userId: string) => {
    try {
        const response = await axiosInstance.get(`/api/cartitem/all?UserId=${userId}`);

        if (response.status >= 200 && response.status < 300) {
            return response.data;
        }

        return {
            success: false,
            message: response.data.message || "Unexpected Error Occurred",
            validationErrors: []
        };
    } catch (error) {
        return {
            success: false,
            message: "Unexpected Error Occurred",
            validationErrors: []
        };
    }
}

const cartService = {
    addItemToCart,
    removeItemFromCart,
    clearCart,
    updateItemInCart,
    getItemsInCart
}

export default cartService;